From 0847bc6e4f0dc3d4f54132146abc690fd2c80036 Mon Sep 17 00:00:00 2001
From: Rik van Riel <riel@redhat.com>
Date: Fri, 5 Sep 2014 14:16:57 -0400
Subject: [kernel] sched: change thread_group_cputime() to use for_each_thread()

Message-id: <1409926623-25929-2-git-send-email-riel@redhat.com>
Patchwork-id: 91453
O-Subject: [RHEL7 PATCH 1/7] sched: change thread_group_cputime() to use for_each_thread()
Bugzilla: 1120307
RH-Acked-by: Larry Woodman <lwoodman@redhat.com>
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Stanislaw Gruszka <sgruszka@redhat.com>

From: Oleg Nesterov <oleg@redhat.com>

Fixes bug 1120307

Upstream status: in peterz's tip queue

Commit f6d1c0103e5a424f1aaee43da39800691d095a5b
Author: Oleg Nesterov <oleg@redhat.com>
Date:   Wed, 13 Aug 2014 21:20:00 +0200

sched: change thread_group_cputime() to use for_each_thread()

Change thread_group_cputime() to use for_each_thread() instead of
buggy while_each_thread(). This also makes the pid_alive() check
unnecessary.

Cc: Mike Galbraith <umgwanakikbuti@gmail.com>
Cc: Hidetoshi Seto <seto.hidetoshi@jp.fujitsu.com>
Cc: Frank Mayhar <fmayhar@google.com>
Cc: Frederic Weisbecker <fweisbec@redhat.com>
Cc: Andrew Morton <akpm@linux-foundation.org>
Cc: Sanjay Rao <srao@redhat.com>
Cc: Larry Woodman <lwoodman@redhat.com>
Cc: Rik van Riel <riel@redhat.com>
Signed-off-by: Oleg Nesterov <oleg@redhat.com>
Signed-off-by: Peter Zijlstra <peterz@infradead.org>
Link: http://lkml.kernel.org/r/20140813192000.GA19327@redhat.com
Signed-off-by: Rik van Riel <riel@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/kernel/sched/cputime.c b/kernel/sched/cputime.c
index c89f3b9..1f4b7f2 100644
--- a/kernel/sched/cputime.c
+++ b/kernel/sched/cputime.c
@@ -294,18 +294,12 @@ void thread_group_cputime(struct task_struct *tsk, struct task_cputime *times)
  times->sum_exec_runtime = sig->sum_sched_runtime;
 
  rcu_read_lock();
- /* make sure we can trust tsk->thread_group list */
- if (!likely(pid_alive(tsk)))
-  goto out;
-
- t = tsk;
- do {
+ for_each_thread(tsk, t) {
   task_cputime(t, &utime, &stime);
   times->utime += utime;
   times->stime += stime;
   times->sum_exec_runtime += task_sched_runtime(t);
- } while_each_thread(tsk, t);
-out:
+ }
  rcu_read_unlock();
 }
 
-- 
1.7.1